be/src/exec/pipeline/crossjoin/cross_join_context.cpp

```diff
@@ -43,4 +45,67 @@ Status CrossJoinContext::_init_runtime_filter(RuntimeState* state) {
     }
     return Status::OK();
 }
+
+bool CrossJoinContext::finish_probe(int32_t driver_seq, const std::vector<uint8_t>& build_match_flags) {
+    std::lock_guard guard(_join_stage_mutex);
```

Comment:
Adopt a state machine and atomic variables  just like in HashJoin instead of mutex.
Justmaking state transitions nonreversible, the concurrent code shall work correctly

Reply:
That's for different purpose.

In `NestloopJoinLeftOperator`, the build table is shared by all probers, so the `build_match_flag` should be merged by all probers after the probe procedure, and it needs to be protected by a mutex.

The state machine is used at `ProbeOperator`, to control the stage of join, which is similar with `HashJoin`.

Reply:
This scenario is the same to broadcast HashJoin, in which case, the probe operators share a common build operator. As 
 long as build operator works in full-materialization mode,  then a state machine can be introduced to control both build/probe side operators to work correctly.